### PR TITLE
Add codecs for onion messages

### PR DESCRIPTION
**Pull request by thomash-acinq**
But do not decrypt them yet, only add a log for now.

### CODE REVIEW

File: src/commonMain/kotlin/fr/acinq/lightning/wire/RecipientData.kt [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+import fr.acinq.bitcoin.PublicKey
+import fr.acinq.bitcoin.io.ByteArrayInput
+import fr.acinq.bitcoin.io.ByteArrayOutput
+import fr.acinq.bitcoin.io.Input
+import fr.acinq.bitcoin.io.Output
+import kotlinx.serialization.Contextual
+import kotlinx.serialization.Serializable
+
+@Serializable
+sealed class RecipientDataTlv : Tlv {
```
Comment (t-bast) [inline src/commonMain/kotlin/fr/acinq/lightning/wire/RecipientData.kt:13]: Can you put this in a file called `RouteBlinding.kt` inside `wire` and rename it `RouteBlindingEncryptedDataTlv` to match what is in `eclair`? It's best to be as close as possible to `eclair` to make it easy to spot differences and reduce maintenance burden.

File: src/commonMain/kotlin/fr/acinq/lightning/wire/RecipientData.kt [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+        companion object : TlvValueReader<NextBlinding> {
+            const val tag: Long = 8
+            override fun read(input: Input): NextBlinding =
+                NextBlinding(PublicKey(LightningCodecs.bytes(input, 33)))
+        }
+    }
+}
+
+data class RecipientData(val records: TlvStream<RecipientDataTlv>) {
+    val nextnodeId = records.get<RecipientDataTlv.OutgoingNodeId>()?.nodeId
```
Comment (t-bast) [inline src/commonMain/kotlin/fr/acinq/lightning/wire/RecipientData.kt:60]: nit: 
```suggestion
    val nextNodeId = records.get<RecipientDataTlv.OutgoingNodeId>()?.nodeId
```

File: src/commonMain/kotlin/fr/acinq/lightning/crypto/RouteBlinding.kt [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+     * @param sessionKey this node's session key.
+     * @param publicKeys public keys of each node on the route, starting from the introduction point.
+     * @param payloads   payloads that should be encrypted for each node on the route.
+     * @return a blinded route.
+     */
+    fun create(sessionKey: PrivateKey, publicKeys: List<PublicKey>, payloads: List<ByteVector>): BlindedRoute {
+        require(publicKeys.size == payloads.size) { "a payload must be provided for each node in the blinded path" }
+        var e = sessionKey
+        val (blindedHops, blindingKeys) = publicKeys.zip(payloads).map {
+            pair ->
```
Comment (t-bast) [inline src/commonMain/kotlin/fr/acinq/lightning/crypto/RouteBlinding.kt:67]: Indenting is weird here, can you run the formatter?

**Review by t-bast — COMMENTED** [review]
I only had a quick look to check that it matches what's inside eclair, it looks mostly good to me.
This PR needs to import tests though:

- for the route blinding crypto part
- for the codecs (we should use the hard-coded hex data from eclair tests to ensure we're able to encode/decode the same thing as eclair)

File: src/commonMain/kotlin/fr/acinq/lightning/crypto/RouteBlinding.kt (hunk 3 above)
Comment (thomash-acinq) [reply src/commonMain/kotlin/fr/acinq/lightning/crypto/RouteBlinding.kt]: Done

File: src/commonMain/kotlin/fr/acinq/lightning/wire/RecipientData.kt (hunk 1 above)
Comment (thomash-acinq) [reply src/commonMain/kotlin/fr/acinq/lightning/wire/RecipientData.kt]: Done

File: src/commonTest/kotlin/fr/acinq/lightning/wire/RouteBlindingTestsCommon.kt [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+        ByteVector("0121000000000000000000000000000000000000000000000000000000000000000000 04210324653eac434488002cc06bbfb7f10fe18991e35f9fe4302dbea6d2353dc0ab1c") to RouteBlindingEncryptedData(TlvStream(listOf(Padding(ByteVector("000000000000000000000000000000000000000000000000000000000000000000")), OutgoingNodeId(PublicKey(ByteVector("0324653eac434488002cc06bbfb7f10fe18991e35f9fe4302dbea6d2353dc0ab1c")))))),
+        ByteVector("0421027f31ebc5462c1fdce1b737ecff52d37d75dea43ce11c74d25aa297165faa2007 0821031b84c5567b126440995d3ed5aaba0565d71e1834604819ff9c17f5e9d5dd078f") to RouteBlindingEncryptedData(TlvStream(listOf(OutgoingNodeId(PublicKey(ByteVector("027f31ebc5462c1fdce1b737ecff52d37d75dea43ce11c74d25aa297165faa2007"))), NextBlinding(PublicKey(ByteVector("031b84c5567b126440995d3ed5aaba0565d71e1834604819ff9c17f5e9d5dd078f")))))),
+        )
+
+        for (payload in payloads) {
+            val encoded = payload.key
+            val data = payload.value
+            println(encoded)
```
[the remark this answers is not in the document]
Comment (thomash-acinq) [reply src/commonTest/kotlin/fr/acinq/lightning/wire/RouteBlindingTestsCommon.kt]: Done

File: src/commonTest/kotlin/fr/acinq/lightning/wire/RouteBlindingTestsCommon.kt [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+        ByteVector("010f000000000000000000000000000000 061000112233445566778899aabbccddeeff") to RouteBlindingEncryptedData(TlvStream(listOf(Padding(ByteVector("000000000000000000000000000000")), PathId(ByteVector("00112233445566778899aabbccddeeff"))))),
+        ByteVector("0121000000000000000000000000000000000000000000000000000000000000000000 04210324653eac434488002cc06bbfb7f10fe18991e35f9fe4302dbea6d2353dc0ab1c") to RouteBlindingEncryptedData(TlvStream(listOf(Padding(ByteVector("000000000000000000000000000000000000000000000000000000000000000000")), OutgoingNodeId(PublicKey(ByteVector("0324653eac434488002cc06bbfb7f10fe18991e35f9fe4302dbea6d2353dc0ab1c")))))),
+        ByteVector("0421027f31ebc5462c1fdce1b737ecff52d37d75dea43ce11c74d25aa297165faa2007 0821031b84c5567b126440995d3ed5aaba0565d71e1834604819ff9c17f5e9d5dd078f") to RouteBlindingEncryptedData(TlvStream(listOf(OutgoingNodeId(PublicKey(ByteVector("027f31ebc5462c1fdce1b737ecff52d37d75dea43ce11c74d25aa297165faa2007"))), NextBlinding(PublicKey(ByteVector("031b84c5567b126440995d3ed5aaba0565d71e1834604819ff9c17f5e9d5dd078f")))))),
```
[the remark this answers is not in the document]
Comment (thomash-acinq) [reply src/commonTest/kotlin/fr/acinq/lightning/wire/RouteBlindingTestsCommon.kt]: Formatted

**Review by t-bast — COMMENTED** [review]
LGTM, just a style nit and we're good to go

**Review by t-bast — APPROVED** [review]
LGTM :rocket:

**Pull request merged**